File: shape-image-view/src/main/java/com/wkz/shapeimageview/GlideImageLoader.java
package com.wkz.shapeimageview;

import android.content.Context;
import android.graphics.drawable.Drawable;
import android.os.Handler;
import android.os.Looper;
import androidx.annotation.DrawableRes;
import androidx.annotation.Nullable;
import android.widget.ImageView;

import com.bumptech.glide.Glide;
import com.bumptech.glide.RequestBuilder;
import com.bumptech.glide.TransitionOptions;
import com.bumptech.glide.load.DataSource;
import com.bumptech.glide.load.engine.GlideException;
import com.bumptech.glide.load.resource.bitmap.CircleCrop;
import com.bumptech.glide.request.RequestListener;
import com.bumptech.glide.request.RequestOptions;
import com.bumptech.glide.request.target.Target;
import com.wkz.shapeimageview.progress.OnGlideImageViewListener;
import com.wkz.shapeimageview.progress.OnProgressListener;
import com.wkz.shapeimageview.progress.ProgressManager;

import java.lang.ref.WeakReference;

/**
 * 图片加载器
 *
 * @author Administrator
 */
public class GlideImageLoader {

    private static final String HTTP = "http";

    private WeakReference<ImageView> mImageView;
    private Object mImageUrlObj;
    private long mTotalBytes = 0;
    private long mLastBytesRead = 0;
    private boolean mLastStatus = false;
    private Handler mMainThreadHandler;

    private OnProgressListener internalProgressListener;
    private OnGlideImageViewListener onGlideImageViewListener;
    private OnProgressListener onProgressListener;


    public GlideImageLoader(ImageView imageView) {
        mImageView = new WeakReference<>(imageView);
        mMainThreadHandler = new Handler(Looper.getMainLooper());
    }

    public ImageView getImageView() {
        return mImageView != null ? mImageView.get() : null;
    }

    public Context getContext() {
        return getImageView() != null ? getImageView().getContext() : null;
    }

    public String getImageUrl() {
        return mImageUrlObj == null || !(mImageUrlObj instanceof String) ? null : (String) mImageUrlObj;
    }

    public void load(Object obj, int... placeholder) {
        load(obj, requestOptions(placeholder));
    }

    public void load(Object obj, RequestOptions options) {
        if (getContext() == null) {
            return;
        }
        requestBuilder(obj, options).into(getImageView());
    }

    public void load(Object obj, TransitionOptions<?, ? super Drawable> transitionOptions, int... placeholder) {
        if (getContext() == null || transitionOptions == null) {
            return;
        }
        requestBuilder(obj, requestOptions(placeholder)).transition(transitionOptions).into(getImageView());
    }

    public RequestBuilder<Drawable> requestBuilder(Object obj, RequestOptions options) {
        this.mImageUrlObj = obj;
        return Glide.with(getContext())
                .load(obj)
                .apply(options)
                .listener(new RequestListener<Drawable>() {
                    @Override
                    public boolean onLoadFailed(@Nullable GlideException e, Object model, Target<Drawable> target, boolean isFirstResource) {
                        mainThreadCallback(mLastBytesRead, mTotalBytes, true, e);
                        ProgressManager.removeProgressListener(internalProgressListener);
                        return false;
                    }

                    @Override
                    public boolean onResourceReady(Drawable resource, Object model, Target<Drawable> target, DataSource dataSource, boolean isFirstResource) {
                        mainThreadCallback(mLastBytesRead, mTotalBytes, true, null);
                        ProgressManager.removeProgressListener(internalProgressListener);
                        return false;
                    }
                });
    }

    public RequestOptions requestOptions(@DrawableRes int... placeholder) {
        RequestOptions requestOptions = new RequestOptions();
        if (placeholder.length == 0) {
            return requestOptions;
        } else if (placeholder.length == 1) {
            return requestOptions.placeholder(placeholder[0])
                    .error(placeholder[0]);
        } else {
            return requestOptions.placeholder(placeholder[0])
                    .error(placeholder[1]);
        }
    }

    public RequestOptions requestOptions(Drawable... placeholder) {
        RequestOptions requestOptions = new RequestOptions();
        if (placeholder.length == 0) {
            return requestOptions;
        } else if (placeholder.length == 1) {
            return requestOptions.placeholder(placeholder[0])
                    .error(placeholder[0]);
        } else {
            return requestOptions.placeholder(placeholder[0])
                    .error(placeholder[1]);
        }
    }

    public RequestOptions circleRequestOptions(@DrawableRes int... placeholder) {
        return requestOptions(placeholder)
                .transform(new CircleCrop());
    }

    public RequestOptions circleRequestOptions(Drawable... placeholder) {
        return requestOptions(placeholder)
                .transform(new CircleCrop());
    }

    private void addProgressListener() {
        if (getImageUrl() == null) {
            return;
        }
        final String url = getImageUrl();
        if (!url.startsWith(HTTP)) {
            return;
        }

        internalProgressListener = new OnProgressListener() {
            @Override
            public void onProgress(String imageUrl, long bytesRead, long totalBytes, boolean isDone, GlideException exception) {
                if (totalBytes == 0) {
                    return;
                }
                if (!url.equals(imageUrl)) {
                    return;
                }
                if (mLastBytesRead == bytesRead && mLastStatus == isDone) {
                    return;
                }

                mLastBytesRead = bytesRead;
                mTotalBytes = totalBytes;
                mLastStatus = isDone;
                mainThreadCallback(bytesRead, totalBytes, isDone, exception);

                if (isDone) {
                    ProgressManager.removeProgressListener(this);
                }
            }
        };
        ProgressManager.addProgressListener(internalProgressListener);
    }

    private void mainThreadCallback(final long bytesRead, final long totalBytes, final boolean isDone, final GlideException exception) {
        mMainThreadHandler.post(() -> {
            final int percent = (int) ((bytesRead * 1.0f / totalBytes) * 100.0f);
            if (onProgressListener != null) {
                onProgressListener.onProgress((String) mImageUrlObj, bytesRead, totalBytes, isDone, exception);
            }

            if (onGlideImageViewListener != null) {
                onGlideImageViewListener.onProgress(percent, isDone, exception);
            }
        });
    }

    public void setOnGlideImageViewListener(OnGlideImageViewListener onGlideImageViewListener) {
        this.onGlideImageViewListener = onGlideImageViewListener;
        addProgressListener();
    }

    public void setOnProgressListener(OnProgressListener onProgressListener) {
        this.onProgressListener = onProgressListener;
        addProgressListener();
    }
}

File: shape-image-view/src/main/java/com/fphoenixcorneae/shapeimageview/GlideImageLoader.kt
package com.fphoenixcorneae.shapeimageview

import android.content.Context
import android.graphics.drawable.Drawable
import android.os.Handler
import android.os.Looper
import android.widget.ImageView
import androidx.annotation.DrawableRes
import com.bumptech.glide.Glide
import com.bumptech.glide.RequestBuilder
import com.bumptech.glide.TransitionOptions
import com.bumptech.glide.load.DataSource
import com.bumptech.glide.load.engine.GlideException
import com.bumptech.glide.load.resource.bitmap.CircleCrop
import com.bumptech.glide.request.RequestListener
import com.bumptech.glide.request.RequestOptions
import com.bumptech.glide.request.target.Target
import com.fphoenixcorneae.shapeimageview.progress.OnGlideImageViewListener
import com.fphoenixcorneae.shapeimageview.progress.OnProgressListener
import com.fphoenixcorneae.shapeimageview.progress.ProgressManager
import java.lang.ref.WeakReference

/**
 * @desc：图片加载器
 */
class GlideImageLoader(imageView: ImageView) {
    private val mImageView: WeakReference<ImageView>?
    private var mImageUrlObj: Any? = null
    private var mTotalBytes: Long = 0
    private var mLastBytesRead: Long = 0
    private var mLastStatus = false
    private val mMainThreadHandler: Handler
    private var internalProgressListener: OnProgressListener? = null
    private var onGlideImageViewListener: OnGlideImageViewListener? = null
    private var onProgressListener: OnProgressListener? = null
    private val imageView: ImageView?
        get() = mImageView?.get()

    private val context: Context?
        get() = imageView?.context

    private val imageUrl: String?
        get() = if (mImageUrlObj == null || mImageUrlObj !is String) null else mImageUrlObj as String?

    fun load(obj: Any?, vararg placeholder: Int) {
        load(obj, requestOptions(*placeholder))
    }

    fun load(obj: Any?, options: RequestOptions) {
        if (context == null) {
            return
        }
        imageView?.let {
            requestBuilder(obj, options).into(it)
        }
    }

    fun load(
            obj: Any?,
            transitionOptions: TransitionOptions<*, in Drawable?>,
            vararg placeholder: Int
    ) {
        if (context == null) {
            return
        }
        imageView?.let {
            requestBuilder(obj, requestOptions(*placeholder))
                    .transition(transitionOptions)
                    .into(it)
        }
    }

    private fun requestBuilder(
            obj: Any?,
            options: RequestOptions
    ): RequestBuilder<Drawable?> {
        mImageUrlObj = obj
        return Glide.with(context!!)
                .load(obj)
                .skipMemoryCache(false)
                .apply(options)
                .listener(object : RequestListener<Drawable?> {
                    override fun onLoadFailed(
                            e: GlideException?,
                            model: Any?,
                            target: Target<Drawable?>,
                            isFirstResource: Boolean
                    ): Boolean {
                        mainThreadCallback(mLastBytesRead, mTotalBytes, true, e)
                        ProgressManager.removeProgressListener(internalProgressListener)
                        return false
                    }

                    override fun onResourceReady(
                            resource: Drawable?,
                            model: Any?,
                            target: Target<Drawable?>,
                            dataSource: DataSource,
                            isFirstResource: Boolean
                    ): Boolean {
                        mainThreadCallback(mLastBytesRead, mTotalBytes, true, null)
                        ProgressManager.removeProgressListener(internalProgressListener)
                        return false
                    }
                })
    }

    /**
     * @param placeholder if placeholder's size is 1,placeholder is placeholder,is error as well.
     *                    if placeholder's size is 2,placeholder[0] is placeholder,placeholder[1] is error.
     */
    private fun requestOptions(@DrawableRes vararg placeholder: Int): RequestOptions {
        val requestOptions = RequestOptions()
        return when {
            placeholder.isEmpty() -> {
                requestOptions
            }
            placeholder.size == 1 -> {
                requestOptions.placeholder(placeholder[0])
                        .error(placeholder[0])
            }
            else -> {
                requestOptions.placeholder(placeholder[0])
                        .error(placeholder[1])
            }
        }
    }

    /**
     * @param placeholder if placeholder's size is 1,placeholder is placeholder,is error as well.
     *                    if placeholder's size is 2,placeholder[0] is placeholder,placeholder[1] is error.
     */
    private fun requestOptions(vararg placeholder: Drawable?): RequestOptions {
        val requestOptions = RequestOptions()
        return when {
            placeholder.isEmpty() -> {
                requestOptions
            }
            placeholder.size == 1 -> {
                requestOptions.placeholder(placeholder[0])
                        .error(placeholder[0])
            }
            else -> {
                requestOptions.placeholder(placeholder[0])
                        .error(placeholder[1])
            }
        }
    }

    fun circleRequestOptions(@DrawableRes vararg placeholder: Int): RequestOptions {
        return requestOptions(*placeholder)
                .transform(CircleCrop())
    }

    fun circleRequestOptions(vararg placeholder: Drawable?): RequestOptions {
        return requestOptions(*placeholder)
                .transform(CircleCrop())
    }

    private fun addProgressListener() {
        if (imageUrl == null) {
            return
        }
        val url = imageUrl
        if (!url!!.startsWith(HTTP)) {
            return
        }
        internalProgressListener = object : OnProgressListener {
            override fun onProgress(
                    imageUrl: String?,
                    bytesRead: Long,
                    totalBytes: Long,
                    isDone: Boolean,
                    exception: GlideException?
            ) {
                if (totalBytes == 0L) {
                    return
                }
                if (url != imageUrl) {
                    return
                }
                if (mLastBytesRead == bytesRead && mLastStatus == isDone) {
                    return
                }
                mLastBytesRead = bytesRead
                mTotalBytes = totalBytes
                mLastStatus = isDone
                mainThreadCallback(bytesRead, totalBytes, isDone, exception)
                if (isDone) {
                    ProgressManager.removeProgressListener(this)
                }
            }
        }
        ProgressManager.addProgressListener(internalProgressListener)
    }

    private fun mainThreadCallback(
            bytesRead: Long,
            totalBytes: Long,
            isDone: Boolean,
            exception: GlideException?
    ) {
        mMainThreadHandler.post {
            val percent = (bytesRead * 1.0f / totalBytes * 100.0f).toInt()
            if (onProgressListener != null) {
                onProgressListener!!.onProgress(
                        mImageUrlObj as String?,
                        bytesRead,
                        totalBytes,
                        isDone,
                        exception
                )
            }
            if (onGlideImageViewListener != null) {
                onGlideImageViewListener!!.onProgress(percent, isDone, exception)
            }
        }
    }

    fun setOnGlideImageViewListener(onGlideImageViewListener: OnGlideImageViewListener?) {
        this.onGlideImageViewListener = onGlideImageViewListener
        addProgressListener()
    }

    fun setOnProgressListener(onProgressListener: OnProgressListener?) {
        this.onProgressListener = onProgressListener
        addProgressListener()
    }

    companion object {
        private const val HTTP = "http"
    }

    init {
        mImageView = WeakReference(imageView)
        mMainThreadHandler = Handler(Looper.getMainLooper())
    }
}